File: server/src/main/java/com/uten/imp/features/production/plancost/ProductionPlanCost.java
package com.uten.imp.features.production.plancost;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.util.UUID;

/**
 * 生产计划成本 / BOM 展开表（生产管理 · 源 F_PlanCostItem，1,359,875 行 · 按年 RANGE 分区）。
 *
 * <p><b>⚠ FK 陷阱</b>（design §3.3）：{@link #billItemId} → {@code production_plan_items.id}
 * （老库 F_PlanCostItem.BillID → F_PlanItem.ID，<b>不是</b> F_Plan.ID！）。
 * 本 Entity 留 UUID {@link #billItemId} <b>不建</b> {@code @ManyToOne}：跨分区 FK 复杂，仅索引 + 应用层保证。
 *
 * <p><b>分区表 PK = (id, bill_date)</b>，但 JPA 单 {@code @Id} 在 {@link #id} 即可查询
 * （{@link #billDate} 作为普通 {@code @Column}）；写入受限——<b>本期只读保数据</b>所以无妨（design §3.3）。
 *
 * <p>数量族（16 个 NUMERIC(18,4)，全保留触发器游标回写的累计量）：
 * {@link #qty} QTY 总需量、{@link #dqty} DQTY 单套用量、{@link #pqty} PQTY 计划数量、
 * {@link #lqty} LQTY 排产占用、{@link #slqty} SLQTY 本次用量（计算列）、{@link #rqty} RQTY 入库、
 * {@link #orderQty} OrderQTY 已订货、{@link #inQty} INQTY 已收货、{@link #pdrawQty} PDrawQTY 已领料、
 * {@link #owdrawQty} OWDrawQTY 已退料、{@link #pwdrawQty} PWDrawQTY 采购退货、
 * {@link #eoQty} EOQTY 委外订货、{@link #eiQty} EIQTY 委外缴回、{@link #ewQty} EWQTY 委外退回、
 * {@link #mqty} MQTY 多订量、{@link #paQty} PAQTY 已排产。
 *
 * <p><b>【本期后置】</b>所有触发器重算逻辑均不实现（design §4.2）：
 * <ul>
 *   <li>BOM 自动展开（TRI_F_PlanCostItem_Insert）—— 归未来 BOM/MRP 模块</li>
 *   <li>数量级联重算（TRI_F_PlanCostItem_Update：父.QTY×DQTY → 子）—— 归未来成本/MRP 模块</li>
 *   <li>Level 计算（TRI_F_PlanCostItem_Level ≤30）—— 迁移原样保 Level，新建走应用层校验</li>
 *   <li>MRP 需购量公式（View_F_PlanCostItem 三分支 CASE）—— 归未来 MRP 模块</li>
 * </ul>
 *
 * <p>多值溯源 9 个 varchar 已在 DDL 合并为 {@link #sourceDocNo}（前缀化保留类型可识别）。
 */
@Getter
@Setter
@NoArgsConstructor
@Entity
@Table(name = "production_plan_costs")
public class ProductionPlanCost {

    /**
     * 单 {@code @Id}（分区表 PK 实为 (id, bill_date)）。Hibernate 单 @Id 对查询无碍；
     * 本期只读不写入（design §3.3）。
     */
    @Id
    @Column(name = "id", nullable = false, updatable = false)
    private UUID id;

    /** F_PlanCostItem.ID（老库 IDENTITY；分区表 UNIQUE 含 bill_date）。 */
    @Column(name = "legacy_id", nullable = false)
    private Integer legacyId;

    /**
     * ⚠ BillID → production_plan_items.id（不是 plans.id！）。
     * 老库 fkeys.txt：F_PlanCostItem.BillID → F_PlanItem.ID（成本展开行挂在计划明细行下）。
     */
    @Column(name = "bill_item_id", nullable = false)
    private UUID billItemId;

    /** 反冗余（JOIN 三级链取，裁剪 + 报表免 JOIN 主表）。 */
    @Column(name = "bill_no", nullable = false)
    private String billNo;

    /** ⭐ 反冗余自 F_Plan.BillDate，分区键（NOT NULL）。 */
    @Column(name = "bill_date", nullable = false)
    private LocalDate billDate;

    /** ParentID 自引用（UUID 映射后回填；不建 FK：分区表 PK 含 bill_date，自引用需含分区键）。 */
    @Column(name = "parent_id")
    private UUID parentId;

    /** 老库 ParentID（0=顶层），迁移留底便于校验。 */
    @Column(name = "parent_legacy_id", nullable = false)
    private Integer parentLegacyId = 0;

    /** Level BOM 层级（TRI_F_PlanCostItem_Level ≤30，原样保）。 */
    @Column(name = "level", nullable = false)
    private Short level = 0;

    /** Class 0=物料 / ≠0=工序或费用。 */
    @Column(name = "node_class", nullable = false)
    private Short nodeClass = 0;

    /** GoodsID 当前节点。 */
    @Column(name = "goods_id", nullable = false)
    private UUID goodsId;

    @Column(name = "color_id")
    private UUID colorId;

    /** MGoodsID 顶层成品（冗余便于按成品汇总）。 */
    @Column(name = "master_goods_id")
    private UUID masterGoodsId;

    @Column(name = "master_color_id")
    private UUID masterColorId;

    /** SOCItemID → sales_order_cost_items.id（V51；跨模块不建 REFERENCES）。 */
    @Column(name = "sales_order_cost_item_id")
    private UUID salesOrderCostItemId;

    // ===== 数量族（16 个，全保留；本期不重算） =====
    @Column(name = "qty", precision = 18, scale = 4)
    private BigDecimal qty = BigDecimal.ZERO;
    @Column(name = "dqty", precision = 18, scale = 4)
    private BigDecimal dqty = BigDecimal.ZERO;
    @Column(name = "pqty", precision = 18, scale = 4)
    private BigDecimal pqty = BigDecimal.ZERO;
    @Column(name = "lqty", precision = 18, scale = 4)
    private BigDecimal lqty = BigDecimal.ZERO;
    @Column(name = "slqty", precision = 18, scale = 4)
    private BigDecimal slqty = BigDecimal.ZERO;
    @Column(name = "rqty", precision = 18, scale = 4)
    private BigDecimal rqty = BigDecimal.ZERO;
    @Column(name = "order_qty", precision = 18, scale = 4)
    private BigDecimal orderQty = BigDecimal.ZERO;
    @Column(name = "in_qty", precision = 18, scale = 4)
    private BigDecimal inQty = BigDecimal.ZERO;
    @Column(name = "pdraw_qty", precision = 18, scale = 4)
    private BigDecimal pdrawQty = BigDecimal.ZERO;
    @Column(name = "owdraw_qty", precision = 18, scale = 4)
    private BigDecimal owdrawQty = BigDecimal.ZERO;
    @Column(name = "pwdraw_qty", precision = 18, scale = 4)
    private BigDecimal pwdrawQty = BigDecimal.ZERO;
    @Column(name = "eo_qty", precision = 18, scale = 4)
    private BigDecimal eoQty = BigDecimal.ZERO;
    @Column(name = "ei_qty", precision = 18, scale = 4)
    private BigDecimal eiQty = BigDecimal.ZERO;
    @Column(name = "ew_qty", precision = 18, scale = 4)
    private BigDecimal ewQty = BigDecimal.ZERO;
    @Column(name = "mqty", precision = 18, scale = 4)
    private BigDecimal mqty = BigDecimal.ZERO;
    @Column(name = "pa_qty", precision = 18, scale = 4)
    private BigDecimal paQty = BigDecimal.ZERO;

    // ===== 金额 =====
    @Column(name = "price", precision = 18, scale = 4)
    private BigDecimal price;

    /** Total = QTY × Price。 */
    @Column(name = "total", precision = 18, scale = 4)
    private BigDecimal total;

    /** VendID 建议供应。 */
    @Column(name = "supplier_id")
    private UUID supplierId;

    /** AssTeamID → B_AssTeam（主档未建，留 legacy int）。 */
    @Column(name = "ass_team_legacy_id")
    private Integer assTeamLegacyId;

    /**
     * 多值溯源号 9 个 varchar 合并文本（前缀化保留类型可识别）：
     * {@code 'PO:xxx | PI:yyy | PW:zzz | PD:aaa | OW:bbb | EO:ccc | EI:ddd | EW:eee | PA:fff'}。
     */
    @Column(name = "source_doc_no")
    private String sourceDocNo;

    /** LStatus 行状态（影响 View_F_PlanCostItem 需购量 CASE 分支）。 */
    @Column(name = "lstatus")
    private Short lstatus;

    /** Summary 摘要。 */
    private String summary;

    // ===== 审计（分区明细省 created_by/updated_by，见 V55） =====
    @Column(name = "created_at", updatable = false)
    private Instant createdAt;

    @Column(name = "updated_at")
    private Instant updatedAt;

    @Column(name = "is_deleted")
    private Boolean isDeleted = false;
}

File: lib/features/production/models/production_plan_cost.dart
// 生产计划成本 / BOM 展开行 model（生产管理 · 只读）。
//
// 对应后端 server/src/main/java/com/uten/imp/features/production/plancost/：
//   GET /production/plan-costs（分页）→ PlanCostRow
//   GET /production/plan-costs/{id} → PlanCostRow
//   GET /production/plan-costs/aggregate → List<PlanCostAggregation>（裸数组）
//
// ⚠ 本期严格只读：F_PlanCostItem 136 万行是"读多写少 + 重算昂贵"的快照表，
//   不做 BOM 展开 / 数量级联重算 / MRP 需购量（归未来成本/MRP 模块，见
//   docs/数据迁移/24 §四、§九）。UI 仅提供查询，标注"只读历史数据"。
//
// 外键陷阱：bill_item_id → production_plan_items.id（不是 plans.id！见 24 §3.3/§五）。
// 分区键：bill_date（按年 RANGE 分区 2018–2030 + DEFAULT）。
//
// JSON：camelCase；16 数量族 NUMERIC(18,4)；sourceDocNo 多值前缀合并
//   'PO:..|PI:..|PW:..|PD:..|OW:..|EO:..|EI:..|EW:..|PA:..'（见 24 §7.3）。

int? _asInt(dynamic v) {
  if (v == null) return null;
  if (v is int) return v;
  if (v is num) return v.toInt();
  if (v is String) return int.tryParse(v);
  return null;
}

double? _asDouble(dynamic v) {
  if (v == null) return null;
  if (v is double) return v;
  if (v is num) return v.toDouble();
  if (v is String) return double.tryParse(v);
  return null;
}

/// BOM 展开行（PlanCostRow，38 字段含 16 数量族 + 单价/金额）。
class ProductionPlanCostRow {
  const ProductionPlanCostRow({
    required this.id,
    this.legacyId,
    this.billItemId,
    this.billNo,
    this.billDate,
    this.parentId,
    this.parentLegacyId,
    this.level,
    this.nodeClass,
    this.goodsId,
    this.colorId,
    this.masterGoodsId,
    this.masterColorId,
    this.salesOrderCostItemId,
    this.qty,
    this.dqty,
    this.pqty,
    this.lqty,
    this.slqty,
    this.rqty,
    this.orderQty,
    this.inQty,
    this.pdrawQty,
    this.owdrawQty,
    this.pwdrawQty,
    this.eoQty,
    this.eiQty,
    this.ewQty,
    this.mqty,
    this.paQty,
    this.price,
    this.total,
    this.supplierId,
    this.assTeamLegacyId,
    this.sourceDocNo,
    this.lstatus,
    this.summary,
  });

  final String id;
  final int? legacyId;
  final String? billItemId; // → production_plan_items.id（不是 plans.id）
  final String? billNo;
  final String? billDate; // 分区键
  final String? parentId; // 自引用（null=顶层）
  final int? parentLegacyId; // 0=顶层
  final int? level; // BOM 层级（≤30）
  final int? nodeClass; // 0=物料 / ≠0=工序费用
  final String? goodsId;
  final String? colorId;
  final String? masterGoodsId; // 顶层成品（冗余便于按成品汇总）
  final String? masterColorId;
  final String? salesOrderCostItemId;

  /// 数量族（16 个，NUMERIC(18,4)；本期不重算，原样保留触发器累计量）。
  final double? qty; // 总需量（成品量×单支用量）
  final double? dqty; // 单套用量（BOM 单耗）
  final double? pqty; // 计划数量（带损耗投产量）
  final double? lqty; // 排产占用
  final double? slqty; // 本次用量（计算列）
  final double? rqty; // 入库数量
  final double? orderQty; // 已订货（采购回写）
  final double? inQty; // 已收货（采购回写）
  final double? pdrawQty; // 已领料（仓库回写）
  final double? owdrawQty; // 已退料（仓库回写）
  final double? pwdrawQty; // 采购退货量
  final double? eoQty; // 委外订货
  final double? eiQty; // 委外缴回
  final double? ewQty; // 委外退回
  final double? mqty; // 多订量（手工调整）
  final double? paQty; // 已排产量

  final double? price;
  final double? total; // = qty × price
  final String? supplierId; // 建议供应
  final int? assTeamLegacyId;
  final String? sourceDocNo; // 多值前缀合并文本
  final int? lstatus;
  final String? summary;

  /// nodeClass 文案：0=物料 / ≠0=工序·费用。
  String get nodeClassLabel {
    switch (nodeClass) {
      case 0:
      case null:
        return '物料';
      default:
        return '工序·费用';
    }
  }

  factory ProductionPlanCostRow.fromJson(Map<String, dynamic> json) =>
      ProductionPlanCostRow(
        id: json['id'] as String,
        legacyId: _asInt(json['legacyId']),
        billItemId: json['billItemId'] as String?,
        billNo: json['billNo'] as String?,
        billDate: json['billDate'] as String?,
        parentId: json['parentId'] as String?,
        parentLegacyId: _asInt(json['parentLegacyId']),
        level: _asInt(json['level']),
        nodeClass: _asInt(json['nodeClass']),
        goodsId: json['goodsId'] as String?,
        colorId: json['colorId'] as String?,
        masterGoodsId: json['masterGoodsId'] as String?,
        masterColorId: json['masterColorId'] as String?,
        salesOrderCostItemId: json['salesOrderCostItemId'] as String?,
        qty: _asDouble(json['qty']),
        dqty: _asDouble(json['dqty']),
        pqty: _asDouble(json['pqty']),
        lqty: _asDouble(json['lqty']),
        slqty: _asDouble(json['slqty']),
        rqty: _asDouble(json['rqty']),
        orderQty: _asDouble(json['orderQty']),
        inQty: _asDouble(json['inQty']),
        pdrawQty: _asDouble(json['pdrawQty']),
        owdrawQty: _asDouble(json['owdrawQty']),
        pwdrawQty: _asDouble(json['pwdrawQty']),
        eoQty: _asDouble(json['eoQty']),
        eiQty: _asDouble(json['eiQty']),
        ewQty: _asDouble(json['ewQty']),
        mqty: _asDouble(json['mqty']),
        paQty: _asDouble(json['paQty']),
        price: _asDouble(json['price']),
        total: _asDouble(json['total']),
        supplierId: json['supplierId'] as String?,
        assTeamLegacyId: _asInt(json['assTeamLegacyId']),
        sourceDocNo: json['sourceDocNo'] as String?,
        lstatus: _asInt(json['lstatus']),
        summary: json['summary'] as String?,
      );
}

/// BOM 成本汇总行（按顶层成品 / 节点货品上卷；GET .../aggregate 裸数组）。
class ProductionPlanCostAggregation {
  const ProductionPlanCostAggregation({
    this.masterGoodsId,
    this.goodsId,
    this.qtySum,
    this.pqtySum,
    this.totalSum,
    this.lineCnt,
  });

  final String? masterGoodsId;
  final String? goodsId;
  final double? qtySum;
  final double? pqtySum;
  final double? totalSum;
  final int? lineCnt;

  factory ProductionPlanCostAggregation.fromJson(Map<String, dynamic> json) =>
      ProductionPlanCostAggregation(
        masterGoodsId: json['masterGoodsId'] as String?,
        goodsId: json['goodsId'] as String?,
        qtySum: _asDouble(json['qtySum']),
        pqtySum: _asDouble(json['pqtySum']),
        totalSum: _asDouble(json['totalSum']),
        lineCnt: _asInt(json['lineCnt']),
      );
}
